pkg/cmd/cli/cmd/rsync/rsyncargs.go

```diff
@@ -0,0 +1,196 @@
+func (a *RsyncArgs) Set(value string) error {
+	// Parse the string into an array of arguments
+	args, err := parseArgs(value)
+	if err != nil {
+		return err
+	}
+	// Determine if the arguments are for valid flags
+	err = allowedRsyncArgs.Parse(args)
+	if err != nil {
+		return err
+	}
+	*a = args
+	return nil
+}
+
+// states for argument list parser
+const (
+	normal = iota
+	inArgument
+	inSingleQuote
+	inDoubleQuote
+	inEscape
+)
+
+// simple functions to improve readability of parser
+
+func isSingleQuote(c rune) bool {
+	return c == '\''
+}
+
+func isDoubleQuote(c rune) bool {
+	return c == '"'
+}
+
+func isEscape(c rune) bool {
+	return c == '\\'
+}
+
+func isSpace(c rune) bool {
+	return unicode.IsSpace(c)
+}
+
+// parseArgs uses a simple state machine to parse the arguments in a bash-style commmand line.
+// It will recognize single-quoted and double-quoted strings. Escaped characters are only
+// recognized in double-quoted strings
+// State machine states: normal (starting state), inArgument, inSingleQuote, inDoubleQuote, escapeChar
+// Transitions:
+// normal:
+//    space -> normal
+//    single-quote -> inSingleQuote
+//    double-quote -> inDoubleQuote
+//    escape -> inEscape (nextState = inArgument)
+//    other -> inArgument (accumulate char for arg)
+// inArgument:
+//    space -> normal (save and reset current arg)
+//    single-quote -> inSingleQuote
+//    double-quote -> inDoubleQuote
+//    escape  -> inEscape (nextState = inArgument)
+//    other -> inArgument (acccumulate char for arg)
+// inSingleQuote
+//    space -> inSingleQuote (accumulate char for arg)
+//    single-quote -> inArgument
+//    double-quote -> inSingleQuote (accumulate char for arg)
+//    escape-char -> inSingleQuote (accumulate char for arg)
+//    other -> inSingleQuote (accumulate char for arg)
+// inDoubleQuote
+//    space -> inDoubleQuote (accmulate char for arg)
+//    single-quote -> inDoubleQuote (accmulate char for arg)
+//    double-quote -> inArgument
+//    escape-char -> inEscape (nextState = inDoubleQuote)
+//    other -> inDoubleQuote (accmulate char for arg)
+// inEscape
+//    any -> nextState (accumulate char for arg)
+func parseArgs(str string) ([]string, error) {
+	args := []string{}
+	arg := &bytes.Buffer{}
+	state := normal
+	nextState := normal
+	for _, c := range str {
+		switch state {
+		case normal:
+			switch {
+			case isDoubleQuote(c):
+				state = inDoubleQuote
+			case isSingleQuote(c):
+				state = inSingleQuote
+			case isEscape(c):
+				nextState = inArgument
+				state = inEscape
+			case isSpace(c):
+				// keep same state
+			default:
+				state = inArgument
+				arg.WriteRune(c)
+			}
+		case inArgument:
+			switch {
+			case isDoubleQuote(c):
+				state = inDoubleQuote
+			case isSingleQuote(c):
+				state = inSingleQuote
+			case isEscape(c):
+				nextState = inArgument
+				state = inEscape
+			case isSpace(c):
+				state = normal
+				args = append(args, arg.String())
```

Comment:
umm, it's not. if you are inarg and you get a space, then you go back to normal. because you were inarg, the arg that you were building is appended to the array of args.